### PR TITLE
#8636: fix search/insert for variable popover using optional chaining

### LINKED ISSUE #8636: Variable popover doesn't show results when using optional chaining
## Describe the bug

- Variable popover doesn't show results when using optional chaining
- FLUP to: https://github.com/pixiebrix/pixiebrix-extension/issues/8241

## To Reproduce

1. In a variable field
2. Type `@input?.`

## Actual behavior

Variable popover shows no results

## Expected behavior

Variable popover expands the `@input` variable

### PR DESCRIPTION
**Pull request by twschiller**
## What does this PR do?

- Closes #8636
- Follow up to https://github.com/pixiebrix/pixiebrix-extension/pull/8635
- Ignores `?` during filtering
- Attempts to preserve existing `?` when a value is selected from the popover

## Discussion

- Eventually we probably want to do real lexing/parsing to generate an AST that can also be used to re-generate the expression

## Future Work

- https://github.com/pixiebrix/pixiebrix-extension/issues/8638

## Checklist

- [x] Add jest or playwright tests and/or storybook stories
- [x] Designate a primary reviewer: @fungairino

### CODE REVIEW

File: src/runtime/pathHelpers.ts [hunk 1]
```diff
@@ -160,18 +206,25 @@ export function getFieldNamesFromPathString(
...
-export function getPathFromArray(parts: Array<string | number>): string {
-  return parts
-    .map((part, index) => {
-      if (part === "" || (typeof part === "string" && /[ .]/.test(part))) {
-        return `["${part}"]`;
-      }
-
-      if (typeof part === "number" || /^\d+$/.test(part)) {
-        return `[${part}]`;
+export function getPathFromArray(
```
Comment (fungairino) [inline src/runtime/pathHelpers.ts:209]: nit: update jsdoc

File: src/runtime/pathHelpers.ts [hunk 2]
```diff
@@ -151,6 +151,52 @@ export function getFieldNamesFromPathString(
...
 
+/**
+ * Normalize an array of parts into an explicit format simpler for joining the path back together.
+ */
+function normalizePart(
+  partOrRecord:
+    | string
+    | number
+    | { part: string | number; isOptional: boolean },
+): { part: string; isOptional: boolean; isBrackets: boolean } {
```
Comment (fungairino) [inline src/runtime/pathHelpers.ts:162]: nit: this would be a useful type  to refactor and name to reuse in other places.

File: src/runtime/pathHelpers.test.ts [hunk 3]
```diff
@@ -145,35 +145,54 @@ describe("getFieldNamesFromPathString", () => {
...
+    expectMatch(["users", ""], 'users[""]');
+    expectMatch(["names", "Dante Alighieri"], 'names["Dante Alighieri"]');
+    expectMatch(
+      ["Ugo Foscolo", "User Location"],
+      '["Ugo Foscolo"]["User Location"]',
+    );
+    expectMatch(["User List", 100, "id"], '["User List"][100].id');
+    expectMatch(
+      ["User List", 100_000_000, "The name"],
+      '["User List"][100000000]["The name"]',
```
Comment (fungairino) [inline src/runtime/pathHelpers.test.ts:176]: Tests that verify a wide range of inputs are better suited to using the built-in `jest.each`. This improves the test output and makes it so that if one expect fails, all others still run.

File: src/components/fields/schemaFields/widgets/varPopup/menuFilters.ts [hunk 4]
```diff
@@ -241,12 +244,7 @@ export function defaultMenuOption(
     return null;
   }
 
-  if (
-    likelyVariable == null ||
-    likelyVariable === "@" ||
-    !likelyVariable.startsWith("@") ||
-    toPath(likelyVariable).length === 0
-  ) {
+  if (isTextOrNull(likelyVariable) || toVarPath(likelyVariable).length === 0) {
```
Comment (fungairino) [inline src/components/fields/schemaFields/widgets/varPopup/menuFilters.ts:247]: nit: reuse the parts variable?
```suggestion
  if (isTextOrNull(likelyVariable) || parts.length === 0) {
```

File: src/components/fields/schemaFields/widgets/varPopup/menuFilters.ts [hunk 5]
```diff
@@ -31,6 +31,21 @@ import { getIn } from "formik";
  */
 export type MenuOptions = Array<[string, UnknownRecord]>;
 
+/**
+ * Returns true if the value is null or is likely plain text/not a variable.
+ */
+function isTextOrNull(value: string | null): boolean {
```
Comment (fungairino) [inline src/components/fields/schemaFields/widgets/varPopup/menuFilters.ts:37]: Since this is specific to variable strings, I'd rename this to something like `isTextOrNullVar`. 

As an aside, if you want to get fancy with TS you can also return a predicate informing the rest of the code that the value is a valid Var. Ex:
```
type Var = `@${NonEmptyString}`;
const a: Var = "@a";

function isTextOrNullVar(value:string | null): value is Var {
```

At least this would be nice if defining `NonEmptyString` weren't so hard in TS:
https://stackoverflow.com/questions/46253340/how-to-write-a-string-type-that-does-not-contain-an-empty-string-in-typescript

Best we can do for string template types is use something like:
```
type Character = 'a' | 'b' | 'c' | ...;
type NonEmptyString = `${Character}${string}`;
```
Which I thought type-fest would have, but it doesn't.

File: src/components/fields/schemaFields/widgets/varPopup/likelyVariableUtils.test.ts [hunk 6]
```diff
@@ -314,3 +315,33 @@ describe("replaceLikelyVariable", () => {
     expect(newCursorPosition).toEqual(endOfVariableIndex);
   });
 });
+
+describe("getFullVariableName", () => {
```
Comment (fungairino) [inline src/components/fields/schemaFields/widgets/varPopup/likelyVariableUtils.test.ts:319]: nit: you might want to try tests with `?` in some unexpected locations, like in the middle or beginning of a part.

**Review by fungairino — APPROVED** [review]

File: src/runtime/pathHelpers.test.ts (hunk 3 above)
Comment (twschiller) [reply src/runtime/pathHelpers.test.ts:176]: Generally agree, was keeping existing style. You'd need to use a table format since the expected value differs per test

**Comment by twschiller** [comment]
@fungairino I discovered another corner case (see failing test). I'm going to look at generating an AST vs. trying to play whack-a-mole with regex rules a bit later this week

File: src/components/fields/schemaFields/widgets/varPopup/menuFilters.ts (hunk 5 above)
Comment (twschiller) [reply src/components/fields/schemaFields/widgets/varPopup/menuFilters.ts:37]: 👍  I'm going to punt on the type -- it doesn't add much value relative to effort given the method is only applicable in the context of the module

**Comment by twschiller** [comment]
Update: I created a separate issue for the corner case: https://github.com/pixiebrix/pixiebrix-extension/issues/8638
